docs-chatbot/generate-embeddings.ts

```diff
@@ -0,0 +1,123 @@
+import dotenv from "dotenv";
+import { Configuration, OpenAIApi } from "openai";
+import { promises as fs } from "fs";
+import { join } from "path";
+import { encode } from "gpt-tokenizer";
+import * as edgedb from "edgedb";
+import e from "dbschema/edgeql-js";
+
+dotenv.config({ path: ".env.local" });
+
+interface Section {
+  id?: string;
+  path: string;
+  tokens: number;
+  content: string;
+  embedding: number[];
+}
+
+async function walk(dir: string): Promise<string[]> {
+  const immediateFiles = await fs.readdir(dir);
+
+  const recursiveFiles: string[][] = await Promise.all(
+    immediateFiles.map(async (file: any) => {
+      const path = join(dir, file);
+      const stats = await fs.stat(path);
+      if (stats.isDirectory()) return walk(path);
+      else if (stats.isFile()) return [path];
+      else return [];
+    })
+  );
+
+  const flattenedFiles: string[] = recursiveFiles.reduce(
+    (all, folderContents) => all.concat(folderContents),
+    []
+  );
+
+  return flattenedFiles.sort((a, b) => a.localeCompare(b));
+}
+
+async function prepareSectionsData(
+  sectionPaths: string[],
+  openai: OpenAIApi
```

Comment:
I like to organise functions to do one thing, OpenAI configuration and initialization is kinda separate from preparing sections data, and `storeEmbeddings`  is a main function that checks the OPEAN_AI key, does initialization of OpenAI and EdgeDB clients and then calls other functions. 

@jaclarke what do you think is cleaner way of organising code here?

Reply:
Maybe split out OpenAI configuration and initialization into its own function that is called from `prepareSectionsData` to return the already configured client?

```js
async function prepareSectionsData(
  sectionPaths: string[]
) {
  const openai = initOpenAIClient();
  …
```

Something like this maybe? It's a bit more complicated to read than initializing the client inside this function, but I think it's easier to understand than passing the client in after initializing it in a function that doesn't use it.

I don't think leaving it in `storeEmbeddings` gets you closer to a function that does one thing. It was already in a function that did more than one thing. I'd just like to get it out of a function that does nothing with it but pass it to another, so it will be easier for people to make sense of when they're going through the tutorial.

Reply:
Actually it looks like you don't use the openai client in the `generate-answer` endpoint, and instead use `fetch`? Is there a reason?

Reply:
@jclarke openai apis doesn't work with edge runtime (because they use axios), there is another library openai-edge that works but I kind of decided to just show "no-sugar" fetch calls... do u think it would be better to use openai-edge? 
I believe in that case openai- can then be used anywhere but we would need to explain why we use openai-edge instead of openai for the embeddings generation.

Reply:
It looks like v4 of the openai library was just released 5 days ago, which apparently supports edge runtime.

Reply:
wow, that's great